feat(workflow_engine): Add DataConditions

saponifi3d opened the pull request:
# Description
[📜 Technical spec]( https://www.notion.so/sentry/Alerts-Create-Issues-9114a498098143178839d584c293fe75?pvs=4#2910558fcd2941b9af88d86fa37fa390)

### PR Changes
- Adds DataConditions
- Connects `DataCondition` with a 1 to many relationship for `DataConditionGroups` the Group model allows us to define `ANY` and `ALL`, and group the conditions for evaluation.

saponifi3d committed:
WIP

saponifi3d committed:
created the DataCondition model

saponifi3d committed:
create relationships for the data condition model

saponifi3d committed:
manual updates for the fixtures / comparators / factories

saponifi3d committed:
Fix merge conflict issue where return type was removed

markstory reviewed (approved):
Schema looks good to me.

saponifi3d committed:
move the link to detectors / workflows off of the data condition, created a new model called data condition group which will also store logic gates for the data conditions

saponifi3d commented on src/sentry/workflow_engine/models/data_condition.py:
```diff
@@ -0,0 +1,26 @@
+    condition_result = models.JSONField()
+    type = models.CharField(max_length=200)
+
+    group = models.ForeignKey(
```
not a huge fan of `group` here, any other recommendations / suggestions? (I'm nervous about name collisions with `Group` for issues.

markstory replied:
What about `condition_group` or `condition_clause` 🤷

saponifi3d replied:
👍 i'll go with `condition_group` i feel like that's different enough from `issues` but close to this. thanks!

wedamija commented on src/sentry/workflow_engine/models/data_condition_group.py:
```diff
@@ -0,0 +1,28 @@
+@region_silo_model
+class DataConditionGroup(DefaultFieldsModel):
+    """
+    A data condition is a way to specify a condition that must be met for a workflow to execute.
```
`A data group is a way to specify a group of conditions that must be met for a workflow action to execute`

wedamija commented on src/sentry/workflow_engine/models/data_condition_group.py:
```diff
@@ -0,0 +1,28 @@
+    logic_type = models.CharField(max_length=200, choices=Type.choices, default=Type.ANY)
+
+    detector = models.ForeignKey(Detector, on_delete=models.CASCADE, blank=True, null=True)
+    workflow_action = models.ForeignKey(
+        WorkflowAction, on_delete=models.CASCADE, blank=True, null=True
+    )
```
Shouldn't we allow multiple workflow actions to be linked here? So this group of conditions fires this group of actions?

saponifi3d committed:
Updates to the models based on discussions throughout the week.
- Changed workflow_action to be action, and more generic. will use this
  to emit events on detectors or send notifications in workflows
- Created lookup tables; detector_workflow and
  workflow_data_condition_group
- Updated detector and workflow models to represent new design

saponifi3d committed:
export new classes in init file

saponifi3d committed:
Add data_condition_group_actions

saponifi3d committed:
dont need a relationship here, using a lookup table instead

saponifi3d committed:
-s on file name

saponifi3d committed:
fix typo in import

saponifi3d committed:
shakes fist at copilot

saponifi3d committed:
remaining model changes to get sql to generate corretly

saponifi3d commented on src/sentry/workflow_engine/models/data_condition.py:
```diff
@@ -0,0 +1,27 @@
+
+    condition = models.CharField(max_length=200)
+    threshold = models.FloatField()
+    condition_result = models.JSONField()
```
This schema would just be a primitive value; either a string for the detector or boolean for the workflow actions. i can imagine some other scenarios that having this be flexible as a json primitive - but it should never be a compex object / array etc.

is there a good way to enforce that? (off the top of my head, we could have a pre-save signal that validates it's a primitive or throws an exception? 🤔)

wedamija replied:
I think we need a json schema to validate it

saponifi3d committed:
Fix the repr a little

wedamija commented on src/sentry/backup/comparators.py:
```diff
@@ -865,6 +865,7 @@ def get_default_comparators() -> dict[str, list[JSONScrubbingComparator]]:
             ],
             "sentry.userrole": [DateUpdatedComparator("date_updated")],
             "sentry.userroleuser": [DateUpdatedComparator("date_updated")],
+            "workflow_engine.datacondition": [DateUpdatedComparator("date_updated", "date_added")],
```
You'll need to add all the new models here

saponifi3d replied:
👍 thanks - yeah, i was avoiding doing all those changes initially just cause of all the things. updated! (hopefully the tests are green now too 🤞)

wedamija commented on src/sentry/workflow_engine/models/workflow.py:
```diff
@@ -15,6 +17,9 @@ class Workflow(DefaultFieldsModel):
     organization = FlexibleForeignKey("sentry.Organization")
 
+    # Required as the 'when' condition for the workflow
+    condition_group = FlexibleForeignKey(DataConditionGroup, blank=True, null=True)
```
Wondering if we should name it as `when_condition_group` or something

saponifi3d replied:
we def can, one thing i liked about keeping this generic is that we can match the same "required" condition group with detectors if we don't have `when`. I thought it was kinda nice to have those two terms linked because this will be responding to what the detectors emit.

wedamija reviewed (commented):
Looks good although I'd like to see the generated sql after ci is fixed

saponifi3d committed:
update factories

saponifi3d committed:
update comprators

saponifi3d committed:
Updated the backups, i think that's a complete model of a simple interaction

saponifi3d committed:
update backups to work

saponifi3d committed:
update the data condition to have 'comparison' field as well, for string comparisons

saponifi3d committed:
Re-introduce the workflow_action model to get CI green, will remove in another PR

saponifi3d committed:
re-add the model to the import so makemigrations doesnt delete the model

saponifi3d committed:
make sure the data condition group action is set correctly

saponifi3d committed:
add the workflowaction to the org relocation scope to avoid test failures until we can remove the model

wedamija commented on src/sentry/workflow_engine/models/action.py:
```diff
@@ -0,0 +1,24 @@
+@region_silo_model
+class Action(DefaultFieldsModel):
+    """
+    Actions are actions that can be taken if the conditions of a DataConditionGrou are satisfied.
```
Nit: `DataConditionGroup`

wedamija commented on src/sentry/workflow_engine/models/action.py:
```diff
@@ -0,0 +1,24 @@
+        Notification = "SendNotificationAction"
+        TriggerWorkflow = "TriggerWorkflowAction"
+
+    required = models.BooleanField(default=False)
```
I've completely forgotten what `required` means. Might be worth a comment

wedamija commented on src/sentry/workflow_engine/models/action.py:
```diff
@@ -0,0 +1,24 @@
+    __relocation_scope__ = RelocationScope.Organization
+    __repr__ = sane_repr("workflow_id", "type")
+
+    # TODO (@saponifi3d): Don't hardcode these values
+    class Type(models.TextChoices):
+        Notification = "SendNotificationAction"
+        TriggerWorkflow = "TriggerWorkflowAction"
```
Will these be full paths to the actions in practice, similar to issue rules?

saponifi3d replied:
That part is a bit unclear to me, i think it'd be cool if we could derive the information from an `init` file or something and automate it as much as possible. if we did it that way w/o the full path, i think we'd need a registry underneath it to link this type to the class itself.

If we did the full id, could likely omit that registry? (i believe we could just get the module by the module id, just want to confirm)

wedamija replied:
Yeah, either way can work out ok. I guess one thing that's nice about a registry is that it's less susceptible to accidental breakage when people move things around.

wedamija commented on src/sentry/workflow_engine/models/data_condition.py:
```diff
@@ -0,0 +1,28 @@
+    __relocation_scope__ = RelocationScope.Organization
+    __repr__ = sane_repr("type", "condition")
+
+    condition = models.CharField(max_length=200)
+    threshold = models.FloatField(blank=True, null=True)
+    comparison = models.CharField(max_length=200, blank=True, null=True)
+
```
Wondering if it's worthwhile to have comments here explaining what the values look like.

wedamija commented on src/sentry/workflow_engine/models/data_condition.py:
```diff
@@ -0,0 +1,28 @@
+    __repr__ = sane_repr("type", "condition")
+
+    condition = models.CharField(max_length=200)
+    threshold = models.FloatField(blank=True, null=True)
```
An example from the doc that `threshold` could be a string value
```
{
	"threshold": "['tag-a', 'tag-b']",
	"condition": "one_of",
	"condition_result": IssuePriority.LOW
	"type": WorkflowActionTagFilter
}
```

Although I'm not sure if it's up to date, since it doesn't include the `comparison` value either.

saponifi3d replied:
Yeah - i added `comparison` to be the string representation and `threshold` to be the value. i found that when i was adding it to the organization and i wanted to setup a detector - i found i needed to have a string comparison, but only had an float.

alternative could be that we store this in 1 field, but we'd have to parse that into the different types based on the condition type.

any recommendations? (i'm missing JS where i can just set the type to be what i want 🤣)

wedamija replied:
I think either option is fine. I guess that one field that is json with an enforced schema is more consistent, since you're going to need to use json to encode the comparison field either way. But I don't really have a strong opinion on this.

saponifi3d committed:
that one time, at band camp... i forgot **kwargs.

wedamija commented on src/sentry/workflow_engine/models/detector.py:
```diff
@@ -15,6 +15,13 @@ class Detector(DefaultFieldsModel, OwnerModel):
     name = models.CharField(max_length=200)
     data_sources = models.ManyToManyField("workflow_engine.DataSource", through=DataSourceDetector)
 
+    condition_group = FlexibleForeignKey(
```
Wondering if this should be named like `workflow_condition_group` or maybe have a comment explaining how it's used?

saponifi3d committed:
Add a comment about what required actions are, they're basically system level events

saponifi3d committed:
Add comments about each field in the data_condition class

saponifi3d committed:
SQL Changes - Update the name of the condition groups on workflows and detectors to be more descriptive

saponifi3d committed:
rebase - comparators

saponifi3d committed:
Improve the comment a bit more for action.required

saponifi3d committed:
update condition_group to be workflow_condition_group when creating the organization

saponifi3d committed:
simple json schema validation

wedamija commented on src/sentry/workflow_engine/models/data_condition.py:
```diff
@@ -0,0 +1,41 @@
+    comparison = models.CharField(max_length=200, blank=True, null=True)
+
+    # The condition_result is the value that is returned if the condition is met, this must be a primitive value
+    condition_result = models.JSONField(validators=[validate_primitive_value])
```
Unfortunately validators aren't run on model save, and they only work with Django forms, which we don't use.

You'd need to have some way to trigger this on save. It might be possible to do this on presave signal, maybe you can manually trigger the validation logic that way. If we wanted that, we could build it into the base class. Possibly opt in to the behaviour via a class variable or on the `Meta` class.

I think this pr is pretty big though, so maybe we're better off figuring out that part in a separate pr.

saponifi3d committed:
remove threshold, change comparison to json

saponifi3d committed:
Add more database constraints

saponifi3d committed:
Add organization to the data condition group to improve exports here

saponifi3d committed:
GENERATED - SQL Migration

saponifi3d committed:
GENERATED - Fixture updates

saponifi3d committed:
GENERATED - Update snapshots

saponifi3d committed:
Exclude the actions from the relocation scope for now, we may want these to be custom, or could just replicate the org id. punting that decision

saponifi3d committed:
Generated - Fixture Updates... again?

saponifi3d commented:
going to take @wedamija 's +1 from yesterday - the changes have just been to fix the relocation scope and get the PR green.
